docs/Dialects/HWArith/RationaleHWArith.md

```diff
@@ -404,10 +404,66 @@ Example:
 %3 = hwarith.div %16, %17 : (si4, ui6) -> si4
 ```
 
+### Casting
+
+Casting between signless and sign-aware types is essential for `HWArith` to
+coexist and interact with core dialects such as `HW` and `Comb`. Defining
+casting rules can be split into two sub-problems: bit-extension and truncation.
+Truncation to a type with identical or smaller bitwidth is simply implemented
+by extracting the least significant bits of the operand. Hence, it is identical
+for sign-aware and signless types. Bit-extension on the other hand needs to
+consider the signedness of the operand to preserve the operand's value. As
+signless types do not express such information by design, no bit-extension can
+be performed. Values of type `si<w>` or `ui<w>` are sign-extended or
+zero-extended, respectively, regardless of the target's signedness or lack
+thereof. Thus, when casting in `HWArith`, first the bitwidth is adjusted to the
+target size, then in a second step the signedness is changed in a bitcast
+manner.
+
+To visualize this behavior study the following example:
+```mlir
+%1 = hwarith.cast %10 : (ui3) -> ui5
+%2 = hwarith.cast %1 : (ui5) -> si5
+// ^-- both casts can be combined into one equivalent cast:
+%0 = hwarith.cast %10 : (ui3) -> si5
+// ^-- this would be lowered to something like this:
+%zero_padding = hw.constant 0 : i2
+%0_lowered = comb.concat %zero_padding, %10 : i2, i3
+```
+
+As an alternative to the proposed *all-in-one* cast op, we considered
+implementing a set of ops with distinct concerns, for example, an op for
+bitcasting between types of the same bitwidth, an op to implementing the
+truncation and two ops for the different bit extension types. Yet, considering
+that `HWArith` is currently intended as short-lived dialect, being immediately
+lowered to `comb`, no obvious benefits emerge. This could also significantly
+increase the code redundancy during IR generation as well as when lowering to
+`comb`.
+
+#### Overview
+
+|    | Input type | Result type            | Behavior                          |
+| -- | :--------- | :--------------------- | :-------------------------------- |
+|(U) | `ui<a>`    | `ui<b>`/`si<b>`/`i<b>` | zero-extension **if** *b* ≥ *a*   |
+|    |            |                        | truncation **if** *b* < *a*       |
+|(S) | `si<a>`    | `ui<b>`/`si<b>`/`i<b>` | sign-extension **if** *b* ≥ *a*   |
+|    |            |                        | truncation **if** *b* < *a*       |
+|(I) | `i<a>`     | `ui<b>`/`si<b>`        | truncation **if** *b* **≤** *a*   |
+|    | `i<a>`     | `ui<b>`/`si<b>`        | prohibited **if** *b* > *a*       |
```

Comment:
Maybe a brief comment on why it is prohibited - I assume it's because extension would imply that the parser magically infers signedness semantics for the value.

Reply:
The summary is good, and the HTML render of the table still looks fine, I guess. But you could also just put the sentence right below the table, maybe with a footnote style reference.

```
†) prohibited because of the ambiguity ...
```

Reply:
Much better, thanks!